src/codecs/bmp/encoder.rs

```diff
@@ -244,12 +244,20 @@ impl<'a, W: Write + 'a> BmpEncoder<'a, W> {
         for row in (0..height).rev() {
             // from the bottom up
             let row_start = row * y_stride;
-            for col in 0..width {
-                let pixel_start = (row_start + (col * x_stride)) as usize;
-                // color value is equal to the palette index
-                self.writer.write_u8(image[pixel_start])?;
-                // alpha is never written as it's not widely supported
+
+            // color value is equal to the palette index
+            if bytes_per_pixel == 1 {
+                // improve performance by writing the whole row at once
+                let row_end = row_start + y_stride;
+                self.writer
+                    .write_all(&image[row_start as usize..row_end as usize])?;
```

Comment:
Mostly comes down to style, but I think this would be slightly nicer as `.write_all(&image[row_start as usize..][..y_stride as usize]`